CONTRIBUTING.md

```diff
@@ -71,6 +75,72 @@ IntelliJ's code generation can be configured to do this automatically:
      Tick:
      - Make generated local variables final
      - Make generated parameters final
+     
+**Null checking of method or function parameters**
+
+If arguments to a constructor set members of the object, or if arguments to a method
+change internal state, e.g. adding an item to a list, they should be checked for null, e.g.
+
+```
+public class MyObject {
+
+    private final String name;
+    private final List<String> myList = new ArrayList<>();
+
+    public MyObject(final String name) {
+        this.name = Objects.requireNonNull(name, "name");
+    }
+
+    public void addItem(final String item) {
+        myList.add(Objects.requireNotNull(item, "item"));
+    }
+}
+
+```
+
+**No TODOs in code**
+
+Please don't leave any TODOs in the code
+
+**New line for each method or constructor parameter**
+
+Methods or constructors should be defined with each parameter on a new line.
+
+```
+public void doSomething(
+    final String name,
+    final int age,
+    final boolean human
+) {
+    // body of method
+}
+
+```
+
+**Private static methods**
+
+Private methods that don't use object state should be declared as static.
+
+**No else after if branch that returns**
+
+If an if branch returns, then omit the else clause:
+
+```
+if (x == 3) {
+    return true;
+}
+return false;
+```
+
+Don't do this:
+
+```
+if (x == 3) {
+    return true;
+} else {
+    return false;
+}
+```
```

Comment:
I would consider adding one about marking all immutable objects with `@Immutable`, and also having a preference for immutability when possible.

Reply:
Not sure whether this counts as a style guideline but Almog's comment reminded me: The use of `@VisibleForTesting` for improved readability

Reply:
Immutable seems reasonable. Tbh I'm not a big fan of @VisibleForTesting, as I don't think it's a good pattern to expose methods on a public API just to allow testing. Another way of doing this is to have a protected method on the API and let a test subclass it to get access to it.

Reply:
we use `@VisibleForTesting` even if its protected/package private (anything other than private)

Reply:
Yeah, we don't have public methods `@VisableForTesting`, but do use it on package level stuff.

Reply:
I think preferring immutable objects is as much a 'style' thing as the null checking, i.e. not letting your objects get into an invalid state.  

It's questionable if its 'style', but we still need to document it somewhere.  Maybe have a 'patterns' section where we document such code patterns?

Oh, and beyond Immutable classes, include the preference for (guava) immutable collections over mutable ones where ever possible.

Reply:
I would put this more under "project recommendations" than rules :)